src/source/Common/EcsCredentialProvider.c

```diff
@@ -0,0 +1,293 @@
+    }
+
+    // Fix-up the expiration to be no more than max enforced token rotation to avoid extra token rotations
+    // as we are caching the returned value which is likely to be an hour but we are enforcing max
+    // rotation to be more frequent.
+    expiration = MIN(expiration, currentTime + MAX_ENFORCED_TOKEN_EXPIRATION_DURATION);
+
+    CHK_STATUS(createAwsCredentials(accessKeyId, accessKeyIdLen, secretKey, secretKeyLen, sessionToken, sessionTokenLen, expiration,
+                                    &pEcsCredentialProvider->pAwsCredentials));
+
+CleanUp:
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS ecsCurlHandler(PEcsCredentialProvider pEcsCredentialProvider)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT64 currentTime;
+    UINT32 formatLen = 0;
+    CHAR serviceUrl[MAX_URI_CHAR_LEN + 1];
+    PRequestInfo pRequestInfo = NULL;
+    CallInfo callInfo;
+
+    MEMSET(&callInfo, 0x00, SIZEOF(CallInfo));
+
+    // Refresh the credentials
+    currentTime = pEcsCredentialProvider->getCurrentTimeFn(pEcsCredentialProvider->customData);
+
+    CHK(pEcsCredentialProvider->pAwsCredentials == NULL ||
+            currentTime + ECS_CREDENTIAL_FETCH_GRACE_PERIOD > pEcsCredentialProvider->pAwsCredentials->expiration,
+        retStatus);
+
+    formatLen = SNPRINTF(serviceUrl, MAX_URI_CHAR_LEN, "%s/%s", pEcsCredentialProvider->ecsGetCredentialEndpoint,
+                         pEcsCredentialProvider->ecsGetCredentialResource);
+    CHK(formatLen > 0 && formatLen < MAX_URI_CHAR_LEN, STATUS_ECS_AUTH_URI_FAILED);
+
+    // Form a new request info based on the params
+    CHK_STATUS(createRequestInfo(serviceUrl, NULL, pEcsCredentialProvider->port, DEFAULT_AWS_REGION, NULL, NULL, NULL, SSL_CERTIFICATE_TYPE_PEM,
+                                 DEFAULT_USER_AGENT_NAME, ECS_REQUEST_CONNECTION_TIMEOUT, ECS_REQUEST_COMPLETION_TIMEOUT, DEFAULT_LOW_SPEED_LIMIT,
+                                 DEFAULT_LOW_SPEED_TIME_LIMIT, pEcsCredentialProvider->pAwsCredentials, &pRequestInfo));
+
+    callInfo.pRequestInfo = pRequestInfo;
+
+    // Append the Ecs header
+    CHK_STATUS(setRequestHeader(pRequestInfo, ECS_AUTH_TOKEN_HEADER, 0, pEcsCredentialProvider->token, 0));
+
+    // Perform a blocking call
+    CHK_STATUS(pEcsCredentialProvider->serviceCallFn(pRequestInfo, &callInfo));
+
+    // Parse the response and get the credentials
+    CHK_STATUS(parseEcsResponse(pEcsCredentialProvider, &callInfo));
+
+CleanUp:
+
+    if (pRequestInfo != NULL) {
+        freeRequestInfo(&pRequestInfo);
+    }
+
+    releaseCallInfo(&callInfo);
+
+    return retStatus;
+}
+
+STATUS getHostPort(PCHAR pUrl, PCHAR* ppStart, PCHAR* ppEnd)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PCHAR pStart = NULL, pEnd = NULL, pCurPtr;
+    UINT32 urlLen;
+    BOOL iterate = TRUE;
+
+    CHK(pUrl != NULL && ppStart != NULL && ppEnd != NULL, STATUS_NULL_ARG);
+
+    // We know for sure url is NULL terminated
+    urlLen = (UINT32) STRLEN(pUrl);
```

Comment:
Some static code analysis tools don't know if it's null terminated :). Recommend using STRNLEN(pUrl, MAX_URL_LEN).

Reply:
Though, it seems that we do have some usage of STRLEN in the codebase.

Reply:
Done